### PR TITLE
feat(db): persist happy-route writer truth

### PR DESCRIPTION
**Pull request by mt4110**
## Summary

Implements issue #21 by replacing the happy-route authoritative in-memory writer path with PostgreSQL-backed writer truth while preserving the current HTTP surface.

## What changed

- Adds forward-only migrations for durable happy-route truth boundaries:
  - core Pi account links and auth sessions
  - Promise intent idempotency keys
  - settlement intents, submissions, provider attempts, and observations
  - raw provider callbacks and raw callback dedupe
  - payment receipt business-key idempotency
  - outbox `manual_review` terminal coordination status
- Replaces the old in-memory happy-route writer with a `tokio-postgres` repository.
- Keeps provider I/O outside authoritative transactions and persists post-I/O results in fresh writer transactions.
- Keeps callback intake raw-evidence-first: the HTTP callback writes raw evidence and outbox work only.
- Preserves `GET /api/projection/settlement-views/{settlement_case_id}` response parity on top of PostgreSQL writer facts.
- Migrates happy-route integration tests to the PostgreSQL test DB and adds direct writer-table assertions for durable facts.
- Updates docs only for the DB truth boundary and validation flow.

## Explicitly out of scope

- Proof persistence
- Refund / release / dispute flows
- Trust read models, expanded Promise read models, ranking, scoring, or broader #22 read-side growth

## Validation

- `make ENV_FILE=.env.example db-bootstrap`
- `make ENV_FILE=.env.example db-migrate`
- `make ENV_FILE=.env.example db-status`
- `cargo check`
- `cargo test`
- `git diff --check`

Note: plain `make db-bootstrap` initially failed because the local untracked `.env` lacks `APP_ENV`; the committed `.env.example` path was used for reproducible validation.

### CODE REVIEW

File: apps/backend/src/services/happy_route/repository.rs [hunk 1]
```diff
@@ -1,11 +1,2476 @@
...
+        realm_id: row.get("settlement_realm_id"),
+        case_status: row.get("case_status"),
+        backend_pin: BackendPin::new(
+            BackendKey::new(backend_key),
+            BackendVersion::new(backend_version),
+        ),
+        created_at: Utc::now(),
+        updated_at: Utc::now(),
+    })
+}
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs:2655]: Good catch. I updated the joined callback-context query to select aliased Promise/settlement timestamps, and the joined-row mappers now hydrate the stored `created_at` / `updated_at` values instead of synthesizing `Utc::now()`.

File: apps/backend/src/services/happy_route/repository.rs [hunk 2]
```diff
@@ -1,11 +1,2476 @@
...
+        amount.currency().as_str(),
+        &amount.scale().to_string(),
+    ]
+    .join("\u{1f}");
+    sha256_hex(material.as_bytes())
+}
+
+fn digest_access_token(access_token: &str) -> String {
+    sha256_hex(access_token.as_bytes())
+}
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs:2861]: Reasonable hardening idea, but I am keeping this PR scoped to the writer-truth migration. This path currently stores a reuse fingerprint for the deferred Pi identity-proof flow; moving to HMAC or a salted verifier needs a separate runtime-secret contract and migration plan.

File: apps/backend/src/services/happy_route/repository.rs [hunk 3]
```diff
@@ -1,11 +1,2476 @@
...
+        PaymentCallbackInput, PaymentCallbackOutcome, PromiseIntentInput, PromiseIntentOutcome,
+        RawPaymentCallbackFields, SettlementViewSnapshot, SubmissionPreparation,
+        processed_outbox_message,
+    },
+};
+
+#[derive(Clone)]
+pub struct HappyRouteStore {
+    client: Arc<Mutex<Client>>,
 }
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs:45]: Fair scaling note, but not for Issue #21. The current DB runtime docs explicitly avoid inventing a pool abstraction before we wire real concurrent writer semantics, so I am leaving the single writer connection in place for now.

File: apps/backend/src/lib.rs [hunk 4]
```diff
@@ -154,6 +213,21 @@ fn env_flag_enabled(name: &str) -> bool {
         .unwrap_or(false)
 }
 
+fn load_env_files() {
+    dotenvy::dotenv().ok();
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/lib.rs:217]: This path is intentional as written: `dotenvy::dotenv()` handles runs from `apps/backend`, and the explicit `apps/backend/.env` fallback handles runs from the repo root. The suggested early-return shape would break the repo-root case.

File: apps/backend/migrations/0007_happy_route_postgres_truth.sql [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+COMMENT ON TABLE core.payment_receipts IS
+'Idempotent MUSUBI payment receipt records derived from verified or reviewable callback evidence. The business key is provider_key plus external_payment_id.';
+
+ALTER TABLE outbox.events
+    DROP CONSTRAINT IF EXISTS outbox_events_delivery_status_check;
+
+ALTER TABLE outbox.events
+    ADD CONSTRAINT outbox_events_delivery_status_check CHECK (
+        delivery_status IN ('pending', 'processing', 'published', 'failed', 'quarantined', 'manual_review')
+    );
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/migrations/0007_happy_route_postgres_truth.sql:180]: Yep, the constraint-name issue is real. I kept the repair as `0008` instead of rewriting `0007`, because we do not want to mutate an already-applied forward migration once the follow-up fix exists in the PR.

File: apps/backend/src/services/happy_route/repository.rs [hunk 6]
```diff
@@ -1,11 +1,2476 @@
...
+            "promise intent id",
+        )?;
+        let raw_callback_id = parse_uuid(&context.raw_callback_id, "raw callback id")?;
+        let amount_minor_units = to_i64(observed_amount.minor_units(), "receipt amount")?;
+        let receipt_status = receipt_status(&verification).to_owned();
+        let verification_observations = verification_observations(&verification)?;
+        let mut client = self.client.lock().await;
+        let tx = client.transaction().await.map_err(db_error)?;
+
+        begin_command_tx(&tx, super::constants::PROVIDER_CALLBACK_CONSUMER, message).await?;
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: The concern is fair, but the suggested code does not match the current `CommandBegin` shape. I kept this PR focused on the concrete correctness fixes above; callback replay already stays side-effect safe through the existing receipt, observation, and ledger idempotency boundaries.

File: apps/backend/src/services/happy_route/repository.rs [hunk 7]
```diff
@@ -1,11 +1,2476 @@
...
+    let row = tx
+        .query_one(
+            "
+            SELECT status
+            FROM outbox.command_inbox
+            WHERE consumer_name = $1
+              AND command_id = $2
+            FOR UPDATE
+            ",
+            &[&consumer_name, &event_id],
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs:1979]: Agreed. `begin_command_tx(...)` now locks and verifies `payload_checksum`, `command_type`, and `schema_version` against the current outbox message before continuing, and I added an integration test that fails on command-inbox drift.

File: apps/backend/src/services/happy_route/repository.rs [hunk 8]
```diff
@@ -1,11 +1,2476 @@
...
+                      AND available_at <= CURRENT_TIMESTAMP
+                    ORDER BY causal_order
+                    LIMIT 1
+                    FOR UPDATE SKIP LOCKED
+                )
+                UPDATE outbox.events events
+                SET delivery_status = $2,
+                    attempt_count = events.attempt_count + 1,
+                    last_attempt_at = CURRENT_TIMESTAMP
+                FROM claimed
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: Agreed. Outbox claiming now writes a lease (`claimed_by` / `claimed_until`), reclaims stale `processing` rows whose lease has expired, and clears the claim fields again on publish/retry/manual-review transitions. I also added an integration test that proves a stale `processing` event is drained successfully.

File: apps/backend/src/services/happy_route/repository.rs [hunk 9]
```diff
@@ -1,11 +1,2476 @@
...
+                .map_err(db_error)?;
+            Some(row.get::<_, Uuid>("first_raw_callback_id"))
+        } else {
+            None
+        };
+
+        let amount_minor_units = raw_fields
+            .and_then(|fields| fields.amount_minor_units)
+            .map(|value| to_i64(value, "callback amount_minor_units"))
+            .transpose()?;
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: Good catch. Callback intake now keeps raw evidence even when `amount_minor_units` does not fit `BIGINT`, and the later drain path preserves that evidence while routing the callback into manual review instead of dropping it. There is an integration test covering the oversized-amount case.

File: apps/backend/tests/happy_route.rs [hunk 10]
```diff
@@ -201,9 +211,226 @@ async fn exact_provider_callback_replay_keeps_the_existing_receipt() {
...
+    let prepared = prepare_pending_case(&app).await;
+
+    let callback = post_json(
+        &app,
+        "/api/payment/callback",
+        None,
+        json!({
+            "payment_id": prepared.payment_id,
+            "payer_pi_uid": prepared.initiator_pi_uid,
+            "amount_minor_units": 9_223_372_036_854_775_808_i128,
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/tests/happy_route.rs]: The “will fail to compile” part is off here: this test already compiled and ran, because the value is still representable in JSON as a positive integer. That said, the current literal is needlessly surprising, so I changed it to the clearer `u64` form.

File: apps/backend/docs/happy_route_walkthrough.md [hunk 11]
```diff
@@ -61,6 +63,8 @@ For `OPEN_HOLD_INTENT` it:
 - emits `REFRESH_SETTLEMENT_VIEW`
 
 In code, that split now starts at `process_open_hold_intent(...)` and is explicitly separated into a prepare write and a post-I/O persistence write so the future PostgreSQL transaction boundary is easier to replace faithfully.
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/docs/happy_route_walkthrough.md]: Agreed. I reworded the walkthrough so it no longer implies the PostgreSQL transaction split is still pending.

File: apps/backend/src/services/happy_route/orchestration.rs [hunk 12]
```diff
@@ -2,29 +2,17 @@ use crate::SharedState;
...
+    types::{DrainOutboxOutcome, HappyRouteError},
 };
 
 pub async fn drain_outbox(state: &SharedState) -> Result<DrainOutboxOutcome, HappyRouteError> {
     let mut processed_messages = Vec::new();
-    {
-        let mut store = state.happy_route.write().await;
-        prune_processed_command_inbox(&mut store);
-    }
+    state.happy_route.prune_processed_command_inbox().await?;
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/orchestration.rs:13]: Agreed. `drain_outbox(...)` now prunes expired terminal outbox rows before claiming new work, and there is an integration test that proves expired published rows are removed from `outbox.events`.

File: apps/backend/src/services/happy_route/repository.rs [hunk 13]
```diff
@@ -1,11 +1,2513 @@
...
+            "promise intent id",
+        )?;
+        let raw_callback_id = parse_uuid(&context.raw_callback_id, "raw callback id")?;
+        let amount_minor_units = to_i64(observed_amount.minor_units(), "receipt amount")?;
+        let receipt_status = receipt_status(&verification).to_owned();
+        let verification_observations = verification_observations(&verification)?;
+        let mut client = self.client.lock().await;
+        let tx = client.transaction().await.map_err(db_error)?;
+
+        begin_command_tx(&tx, super::constants::PROVIDER_CALLBACK_CONSUMER, message).await?;
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: Agreed. `persist_payment_callback_result(...)` now short-circuits when `begin_command_tx(...)` reports `Completed`, marks the outbox event published, and returns the persisted replay outcome instead of appending new observations or reapplying effects. I added an integration test for that completed-command replay path.

File: apps/backend/src/services/happy_route/repository.rs [hunk 14]
```diff
@@ -1,11 +1,2549 @@
...
+        )
+        .await
+        .map_err(db_error)?;
+        tx.execute(
+            "
+            INSERT INTO dao.promise_intent_idempotency_keys (
+                initiator_account_id,
+                internal_idempotency_key,
+                promise_intent_id,
+                request_payload_hash
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: Fixed. The Promise idempotency insert now uses ON CONFLICT DO NOTHING, deletes the losing transaction rows before commit, and reloads the winning idempotency row as a replay instead of surfacing a unique-constraint 500. I also added a blocker-transaction integration test that reproduces the concurrent duplicate-submit race and now returns the original successful result.

File: apps/backend/src/services/happy_route/repository.rs [hunk 15]
```diff
@@ -1,11 +1,2549 @@
...
+            )
+            .await
+            .map_err(db_error)?;
+            tx.execute(
+                "
+                INSERT INTO core.pi_account_links (
+                    account_id,
+                    pi_uid,
+                    username,
+                    wallet_address,
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: Fixed. First-time Pi account-link creation now goes through a race-safe ON CONFLICT (pi_uid) path, deletes the losing temporary account row, and only updates the existing link metadata when the access-token digest matches. I added a blocker-transaction integration test to verify the retry path now reuses the winning account instead of returning a 500.

File: apps/backend/src/services/happy_route/backend.rs [hunk 16]
```diff
@@ -565,6 +534,18 @@ fn classify_callback_status(status: &str) -> CallbackStatusClass {
...
 }
 
+fn store_error_to_backend_error(error: HappyRouteError) -> BackendError {
+    match error {
+        HappyRouteError::NotFound(_) => BackendError::InvalidProviderPayload,
+        HappyRouteError::BadRequest(_) => BackendError::InvalidProviderPayload,
+        HappyRouteError::Unauthorized(_) => BackendError::InvalidProviderPayload,
+        HappyRouteError::ProviderCallbackMappingDeferred(_) => BackendError::TemporarilyUnavailable,
+        HappyRouteError::Provider { .. } | HappyRouteError::Internal(_) => {
+            BackendError::InvalidProviderResponse
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/backend.rs]: Fixed. `store_error_to_backend_error` now treats database-backed `HappyRouteError::Internal` failures as `BackendError::TemporarilyUnavailable`, so transient writer/store failures stay on the retryable path instead of being escalated to terminal manual review. I also added unit coverage for both the database-failure retry case and the fail-closed non-database internal case.

File: apps/backend/src/services/happy_route/repository.rs [hunk 17]
```diff
@@ -1,11 +1,2636 @@
...
+        .execute(
+            "
+            UPDATE outbox.events
+            SET delivery_status = $2,
+                claimed_by = NULL,
+                claimed_until = NULL,
+                last_error_class = $3,
+                last_error_detail = $4,
+                retain_until = NULL
+            WHERE event_id = $1
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: Fixed. Terminal outbox transitions now keep a retention deadline instead of clearing `retain_until`, so quarantined and manual-review rows remain inspectable for a bounded window and still satisfy the prune predicate later. The new malformed-payload integration test also checks that an expired quarantined row is actually removed on the next drain cycle.

File: apps/backend/src/services/happy_route/repository.rs [hunk 18]
```diff
@@ -1,11 +1,2636 @@
...
+                    events.available_at,
+                    events.published_at,
+                    events.created_at
+                ",
+                &[&OUTBOX_PENDING, &OUTBOX_PROCESSING, &"happy-route-drain"],
+            )
+            .await
+            .map_err(db_error)?;
+        tx.commit().await.map_err(db_error)?;
+        row.map(|row| outbox_message_from_row(&row)).transpose()
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: Fixed. `claim_pending_outbox_message` now validates the claimed row inside the transaction, quarantines invalid payloads with terminal diagnostics, preserves a prune deadline, and then continues scanning for the next claimable event. I added an integration test that inserts a malformed outbox row, verifies it is quarantined instead of left stuck in `processing`, and confirms later drains can prune it once the retention window expires.

File: apps/backend/src/services/happy_route/backend.rs [hunk 19]
```diff
@@ -565,6 +534,23 @@ fn classify_callback_status(status: &str) -> CallbackStatusClass {
...
+        HappyRouteError::ProviderCallbackMappingDeferred(_) => BackendError::TemporarilyUnavailable,
+        HappyRouteError::Provider { .. } => BackendError::InvalidProviderResponse,
+        HappyRouteError::Internal(message) => {
+            if message.starts_with("database operation failed:") {
+                BackendError::TemporarilyUnavailable
+            } else {
+                BackendError::InvalidProviderResponse
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/backend.rs:574]: Fixed. I replaced the message-prefix heuristic with a dedicated `HappyRouteError::Database` variant that preserves retryability plus SQLSTATE/constraint metadata. `store_error_to_backend_error` now keeps connection/deadlock-style failures on the retryable path, maps permanent database failures fail-closed, and routes `provider_attempts_provider_reference_key` unique violations to `BackendError::IdempotencyMappingFailed`. I added backend unit coverage for retryable DB failures, permanent DB failures, and the provider-reference uniqueness case.

File: apps/backend/src/services/happy_route/repository.rs [hunk 20]
```diff
@@ -1,11 +1,2678 @@
...
+
+    pub(super) async fn prune_processed_command_inbox(&self) -> Result<(), HappyRouteError> {
+        let client = self.client.lock().await;
+        client
+            .execute(
+                "
+                DELETE FROM outbox.command_inbox
+                WHERE status = 'completed'
+                  AND completed_at IS NOT NULL
+                  AND completed_at < CURRENT_TIMESTAMP - ($1::bigint * interval '1 minute')
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: Fixed. `record_outbox_failure` now finalizes command-inbox state alongside outbox state: retryable failures upsert the inbox row back to `pending`, terminal/manual-review failures upsert it to `quarantined` with retention metadata, and `prune_processed_command_inbox` now prunes expired `completed` and `quarantined` rows by `retain_until`. I added coverage for terminal provider-callback failures so the inbox row is present, unclaimed, and prunable.

File: apps/backend/src/services/happy_route/repository.rs [hunk 21]
```diff
@@ -1,11 +1,2678 @@
...
+    tx.execute(
+        "
+        UPDATE outbox.command_inbox
+        SET status = 'completed',
+            processed_at = CURRENT_TIMESTAMP,
+            completed_at = CURRENT_TIMESTAMP,
+            result_type = $3,
+            result_json = $4,
+            retain_until = CURRENT_TIMESTAMP + ($5::bigint * interval '1 minute')
+        WHERE consumer_name = $1
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs:2034]: Fixed. `complete_command_tx` now clears `claimed_by` and `claimed_until` when a command reaches `completed`, and I added an integration test that verifies completed inbox rows no longer look actively leased after a successful open-hold drain.

File: apps/backend/src/services/happy_route/backend.rs [hunk 22]
```diff
@@ -237,41 +234,22 @@ impl SettlementBackend for PiSettlementBackend {
...
-                    .insert(attempt.provider_attempt_id.clone(), attempt);
+            if existing_attempt.request_hash != attempt.request_hash {
+                return Err(BackendError::IdempotencyMappingFailed);
             }
+        } else {
+            self.state
+                .happy_route
+                .insert_provider_attempt(&attempt)
+                .await
+                .map_err(store_error_to_backend_error)?;
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/backend.rs:265]: Fixed. The backend now proactively checks for an existing attempt by `provider_ref` before insert, and the structured database-error mapping also treats a `provider_attempts_provider_reference_key` unique violation as `BackendError::IdempotencyMappingFailed` rather than a retryable store failure. That closes both the obvious pre-insert collision and the insert-race fallback.

File: apps/backend/src/services/happy_route/repository.rs [hunk 23]
```diff
@@ -1,11 +1,2867 @@
...
+                        message,
+                        "transient",
+                        error.message(),
+                    )
+                    .await?;
+                    tx.execute(
+                        "
+                        UPDATE outbox.events
+                        SET delivery_status = $2,
+                            claimed_by = NULL,
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs:767]: Fixed. record_outbox_failure now treats unclassified/internal coordination errors as terminal and quarantines both outbox.events and the corresponding command-inbox row instead of returning them to pending. I also extended drain_outbox_rejects_mismatched_command_inbox_payloads to assert the corrupted event and inbox rows land in quarantined.

File: apps/backend/src/services/happy_route/callback.rs [hunk 24]
```diff
@@ -83,30 +65,32 @@ pub(super) async fn process_provider_callback(
...
+        if let Some(outcome) = state
+            .happy_route
+            .payment_callback_replay_outcome(&callback_context)
+            .await?
         {
-            super::outbox::mark_outbox_published(&mut store, &message.event_id);
+            state
+                .happy_route
+                .mark_outbox_published(&message.event_id)
+                .await?;
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/callback.rs:93]: Fixed. The duplicate-callback fast path now goes through finalize_provider_callback_replay, which begins or validates the provider-callback command row, completes it when needed, and publishes the outbox event in the same transaction. I extended exact_provider_callback_replay_keeps_the_existing_receipt with a pre-existing pending inbox row and assert it ends completed and unclaimed.

File: apps/backend/src/services/happy_route/repository.rs [hunk 25]
```diff
@@ -1,11 +1,2867 @@
...
+            } else {
+                duplicate_receipt = true;
+                should_apply_verified_effects = false;
+            }
+        } else {
+            tx.execute(
+                "
+                INSERT INTO core.payment_receipts (
+                    payment_receipt_id,
+                    provider_key,
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs]: Fixed. The payment receipt insert now uses ON CONFLICT (provider_key, external_payment_id) DO NOTHING; if another worker wins the insert race, the loser reloads the existing receipt under FOR UPDATE and replays or upgrades based on receipt status instead of surfacing a unique-constraint error. I added concurrent_payment_receipt_insert_replays_race_winner to cover that race.

File: apps/backend/tests/happy_route.rs
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/tests/happy_route.rs:1117]: Fixed. I replaced the fixed sleep with a DB-level lock-contention probe via pg_blocking_pids(...), so the test waits until the spawned request is actually blocked on the open transaction before the blocker commits.

[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/tests/happy_route.rs:1454]: Fixed. The race tests now use the same pg_blocking_pids(...) helper instead of sleep-based timing, so payment-receipt and promise/auth replay coverage waits for real lock contention before releasing the blocker transaction.

File: apps/backend/src/services/happy_route/repository.rs [hunk 26]
```diff
@@ -1,11 +1,2905 @@
...
+        }
+
+        tx.execute(
+            "
+            INSERT INTO projection.settlement_views (
+                settlement_case_id,
+                realm_id,
+                promise_intent_id,
+                latest_journal_entry_id,
+                current_settlement_status,
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/happy_route/repository.rs:1610]: Fixed. get_settlement_view now maps settlement_case_id UUID parse failures to settlement_projection_not_found before touching the projection table, so the HTTP surface stays on the existing 404 path instead of leaking a new 400 shape. I also added settlement_projection_keeps_invalid_ids_as_not_found to pin that contract.
